### ISSUE TITLE
`$schema` should always be an allowed property

### ISSUE THREAD

**Issue by MarkCatherall** [json-schema/json-schema#220]
Hello, I'm loving JSON schema, great work. One little quirk which I've encountered is that if you build a schema for an object which specifies `"additionalProperties": false` then that object cannot have a `"$schema"` property unless you specify `"$schema"` as one of the properties of the object within your schema. This is kind of a catch-22, because of course if you hadn't specified a schema in your object, it wouldn't know that `"$schema"` wasn't allowed!

Would there be any downside to making `"$schema"` an exception to `"additionalProperties": false`?

Thanks again.

Mark

**Comment by epoberezkin** [comment]
`$schema` is supposed to be used in the schema at the top level, not in the data object. It's a reference to the meta-schema (the schema that can be used to validate the schema). There is no spec in the standard that allows to specify inside the data which schema should be used to validate it.

**Comment by MarkCatherall** [comment]
OK, but that use of `$schema` to specify the meta-schema is just an instance of an object specifying the schema that should be used to validate it, i.e., you write a schema and put in the `$schema` keyword to specify which schema should be used to validate it. Alternatively, you write some json and put in the `$schema` keyword to specify which schema should be used to validate it. Why wouldn't this be scale-invariant? This is the way that Visual Studio interprets the `$schema` keyword: https://blogs.msdn.microsoft.com/webdev/2014/04/10/intellisense-for-json-schema-in-the-json-editor/

**Comment by sam-at-github** [comment]
> Would there be any downside to making "$schema" an exception to "additionalProperties": false?

Yeah. currently additionalProperties means exactly that, not additional properties with the exception of ... Plus that would make it hard to say you don't want a "$schema" property. Not a good idea IMO.

**Comment by robbie-mac** [comment]
@MarkCatherall  looking at the article in your link, they are just showing an example. However, in that example, you wont get any of the JSON Schema core intellisense because you just created a circular reference. It will validate that schema against itself. Not against the JSON Schema Core V4. So in that sense you are creating the base schema and are capable of adding $schema to whatever element you want... it just wont be the 'standard' JSON Schema.

**Comment by MarkCatherall** [comment]
@robbie-mac I'm a bit confused by your comment: there are no circular references, schemas that I write are validated against the Core V4 schema, documents that I write are validated against whatever schema I point them to with the `$schema` value, maybe one I've just written, maybe one someone else has written. At no point does a schema point to itself, the only possible case where that could be the right behaviour is that the Core V4 schema might point to itself.

The intellisense that you get when you do this is useful stuff from the schema that you've written and pointed your document to, like the title, description, and numeric restrictions on the value you're hovering over. Try it, it's great!

**Comment by MarkCatherall** [comment]
@robbie-mac, to make it more clear than in Microsoft's example, consider the following:

If I write this schema (called testSchema.json):
```
{
  "$schema": "http://json-schema.org/draft-04/schema",
  "type": "object",
  "title": "Container",
  "description": "Brilliant json file that I've created.",
  "properties": {
    "$schema": {
      "type": "string",
      "title": "Schema",
      "description": "Pointer to the schema against which this document should be validated."
    },
    "numberOfSomethings": {
      "type": "number",
      "title": "Number Of Somethings",
      "minimum": 0,
      "maximum": 100,
      "description": "This number is the number of somethings that there are."
    }
  }
}
```

and this is the document that points to it with the `$schema` value:
```
{
  "$schema": "testSchema.json",
  "numberOfSomethings": 75
}
```

Then this is the intellisense from hovering over the `numberOfSomethings` field:

![image](https://cloud.githubusercontent.com/assets/16152648/14495332/e5c4ad74-0186-11e6-9166-78e9734ac73d.png)

**Comment by robbie-mac** [comment]
@MarkCatherall As I said, in the example in the article you referenced, they (I previously said you, sorry) created a circular reference. By putting the url of the file they are creating. The document they are creating is bieng validated against itself. Circular reference.

I use VS daily and love the intellisense, I was just pointing out that in the example, if you follow their steps you dont get the intellisense from the core schema.

(I'll update this with code when I get to a computer)

In your current example, if you are creating a chain of schemas, why dont you just have 2 schema files and use the ID field to resolve the references.


Also, I may be wrong, but I believe that your data does not get a $schema field. That is one of the things that identifies a document as a schema. Your application should know what schema to validate it against. So in VS case, dont put it in your data document, just use the dropdown at the top of the JSON editor. If you cant find the schema you need, drag your schema file onto the dropdown.

**Comment by MarkCatherall** [comment]
@robbie-mac I'm obviously doing a bad job of explaining myself, maybe I do it better here: https://github.com/spacetelescope/understanding-json-schema/issues/50#issuecomment-196817367

**Comment by robbie-mac** [comment]
Ok, read the other conversation, I get where you are coming from now.
@MarkCatherall 
### TL;DR
Consider this:
Why do we validate data?
We validate data to make sure it conforms to the structure we want/need for our logic to work, we do not process the data for values, simply structure. By attempting to read the $schema from our data, we have just assumed the structure, and are attempting to process the data for values before we know its structure.
Some people say is not as important for data we have created ourselves, only data from unknown source. I say, treat all data as coming from an unknown source, as you don't know what systems have touched, altered, mangled, replaced, or otherwise messed with your data. I know this isn't always possible, but, in those cases (like a JSONSchema ) we test in other ways, before we validate it. We still validate it though.

It all comes down to your level of trust in the data. As an analogy, its like diving off a cliff into water. Wouldn't you like to know how deep the water is before you jump? 
 
### Now to the long version

At the end of the day, a JSON schema is just JSON data written in a particular structure. It does not become a schema until you use it as a schema. Meaning, how it is processed by your application. Same thing with the JSON data you/your application intend as data. So if you want a schema that defines a document with a $schema tag at the root, and peppered throughout your data, its up to you to define that schema, and to make your application make use of that data.

So CAN that data (which may or may not be used as a schema) contain a $schema tag? Absolutely, can contain multiple $schema tags, it could be all $schema tags and nothing else.  It is only restricted by the Schema it is validated against.
In the core-spec schema, the answer is yes, you can specify $schema tags at all levels (It does, and it validates against itself). Except when you tell the validator to NOT let you, like in the situation you mentioned in your original question. If you don't specify $schema as an allowed property and you then tell the validator 'Don't allow any unspecified properties!' by using `"additionalProperties" : false`, the validator wont allow the presence of $schema to pass validation. Why? You told it not to.
You could make it understand what I think you are trying to do, by passing it a list of allowed properties, and tell it to not allow any properties you haven't specified, but also give it a list of properties that are required. If you include $schema in your properties, but not in the required. You have just made $schema an optional property, not an additional property. Your object will pass validation.

To answer your question, yes, I do think there is a downside to making "$schema" an exception to `"additionalProperties":false`. IMO it waters down that rule. Lets keep it concrete, especially since you can achieve the behavior you want already. *The sarcastic reason would be that its a Boolean value, its either true or false. not true or false except... ;)*

As for the other conversation. In respect to the schema, I don't think the spec is ambiguous. It is written for the single use of applying it to data that is a JSONSchema, not data that will be validated against a JSONSchema. End data is out of its scope. IMHO

**Comment by sharwell** [comment]
We are considering the use of JSON Schema to improve the editor experience for one of our formats. The tool which consumes this format does not allow any properties which it does not recognize (strict validation), and `$schema` is not a recognized property. We use `"additionalProperties": false` in the JSON Schema to indicate that you are not allowed to use other properties, so with the current behavior of this property the editor will warn the user if they include a `$schema` property. *This is the desired behavior*, because the tool would reject such an input.

Aside from considering allowing `$schema` in a future release, we have found that JSON editors which support JSON Schema tend to have a second mechanism by which a schema can be associated with the file for editing purposes. `$schema` is convenient and even preferable for users in an editor that recognizes it, but it's not the only way and a change to treat it as special would be problematic for our use case.

**Comment by handrews** [comment]
@MarkCatherall this repository is now deprecated and no one with permissions to do things like close issues is still with the project.  Could you please close this?  If you would like to continue the discussion, please re-open at json-schema-org/json-schema-spec
